addon/background.ts

```diff
@@ -10,18 +11,79 @@ const autoconsentConfig: Config = {
   detectRetries: 20,
 };
 
-async function loadRules(): Promise<RuleBundle> {
+// Storage abstraction: MV2 keeps everything in memory, MV3 uses chrome.storage
+const storage: {[key: string]: any} = {};
+async function storageSet(obj: {[key: string]: any}): Promise<void> {
+  if (manifestVersion === 2) {
+    Object.assign(storage, obj);
+    return;
+  } 
+  return chrome.storage.local.set(obj);
+}
+
+async function storageGet(key: string): Promise<{[key: string]: any}> {
+  if (manifestVersion === 2) {
+    return storage;
+  }
+  return chrome.storage.local.get(key);
+}
+
+async function storageRemove(key: string): Promise<void> {
+  if (manifestVersion === 2) {
+    delete storage[key];
+    return;
+  }
+  return chrome.storage.local.remove(key);
+}
+
+async function loadRules() {
   const res = await fetch("./rules.json");
-  return await res.json();
+  storageSet({
+    rules: await res.json(),
+  });
 }
 
-chrome.runtime.onInstalled.addListener(() => {
-  loadRules().then(rules => {
-    chrome.storage.local.set({
-      rules: rules,
+async function evalInTab(tabId: number, frameId: number, code: string): Promise<chrome.scripting.InjectionResult[]> {
+  if (manifestVersion === 2) {
+    return new Promise((resolve) => {
+      chrome.tabs.executeScript(tabId, {
+        frameId,
+        code: `window.eval(JSON.parse('${JSON.stringify(code)}'))`
```

Comment:
on etsy.com, I get an unexpected `undefined` result. Should we add "return" in the snippet here?
<img width="980" alt="image" src="https://user-images.githubusercontent.com/2726132/179471743-36b8c7ad-6ec6-4a9d-9eab-5d102ca88b5f.png">

Reply:
Good catch - I wasn't escaping quotes properly - I've fixed this now.